libfwupdplugin/fu-firmware.c

```diff
@@ -1038,13 +1099,52 @@ fu_firmware_write(FuFirmware *self, GError **error)
 	if (klass->write != NULL)
 		return klass->write(self, error);
 
-	/* set */
-	if (priv->bytes != NULL)
-		return g_bytes_ref(priv->bytes);
-
 	/* just add default blob */
-	g_set_error_literal(error, FWUPD_ERROR, FWUPD_ERROR_NOT_FOUND, "no payload set");
-	return NULL;
+	return fu_firmware_get_bytes_with_patches(self, error);
+}
+
+/**
+ * fu_firmware_add_patch:
+ * @self: a #FuFirmware
+ * @offset: an address smaller than fu_firmware_get_size()
+ * @blob: (not nullable): bytes to replace
+ *
+ * Adds a byte patch at a specific offset. If a patch already exists at the specified address then
+ * it is replaced.
+ *
+ * If the @address is larger than the size of the image then an error is returned.
+ *
+ * Since: 1.7.4
+ **/
+void
+fu_firmware_add_patch(FuFirmware *self, gsize offset, GBytes *blob)
+{
+	FuFirmwarePrivate *priv = GET_PRIVATE(self);
+	FuFirmwarePatch *ptch;
+
+	g_return_if_fail(FU_IS_FIRMWARE(self));
+	g_return_if_fail(blob != NULL);
+
+	/* ensure exists */
+	if (priv->patches == NULL)
+		priv->patches =
+		    g_ptr_array_new_with_free_func((GDestroyNotify)fu_firmware_patch_free);
+
+	/* find existing */
+	for (guint i = 0; i < priv->patches->len; i++) {
+		ptch = g_ptr_array_index(priv->patches, i);
+		if (ptch->offset == offset) {
```

Comment:
You need to make sure the new patch is the same length or you might be throwing away part of the old one.

Reply:
Absolutely correct, fixed. Do you think we need a `fu_firmware_remove_patch` (just taking the offset) too?

Reply:
I don't know how important that is - users can just re-run to redo patches right?

Reply:
Right, just from a programmatic API point of view.

Reply:
I don't see harm in it if you think something might use it some day.

The only use case I can think of is some vendor wants to ship patches inside a cab and the patches are applied by the software. If that's not a good idea then I would say no need.

Reply:
> The only use case I can think of

Ya, I couldn't think of a use case either; lets add it if/when we need it. Lets see if this works for the reporter.